Se Agrega Metodo Timbrado V4 Stamp

RICH-CD opened the pull request:
Se agrega Timbrado V4 Stamp

RICH-CD committed:
Se Agrega Metodo Timbrado V4 Stamp

JuanGamezSW commented on Test_SW-sdk-45/Services/Stamp/Stamp_Test.cs:
```diff
@@ -13,6 +13,55 @@ namespace Test_SW.Services.Stamp_Test
+            Assert.IsTrue(response.status == "success"
+                && !string.IsNullOrEmpty(response.data.cfdi), "El resultado data.cfdi viene vacio.");
+        }
+        [TestMethod]
```
Agregar UT sin email.

RICH-CD replied:
Se agrega clase StampV4_Test para meter mas TestMethod (UT sin email, UT con customid, UT con email y customid, UT con varios correos)

JuanGamezSW commented on Test_SW-sdk-45/Services/Stamp/Stamp_Test.cs:
```diff
@@ -13,6 +13,55 @@ namespace Test_SW.Services.Stamp_Test
+            Assert.IsTrue(!string.IsNullOrEmpty(response.data.selloCFDI), "El resultado data.selloCFDI viene vacio.");
+            Assert.IsTrue(!string.IsNullOrEmpty(response.data.fechaTimbrado), "El resultado data.fechaTimbrado viene vacio.");
+            Assert.IsTrue(!string.IsNullOrEmpty(response.data.qrCode), "El resultado data.qrCode viene vacio.");
+        }
```
UT con varios email.

RICH-CD replied:
Se agrega clase StampV4_Test para meter mas TestMethod (UT sin email, UT con customid, UT con email y customid, UT con varios correos)

JuanGamezSW commented on SW-sdk/Services/Stamp/StampServiceV4.cs:
```diff
@@ -0,0 +1,37 @@
+        protected StampServiceV4(string url, string token, string proxy, int proxyPort) : base(url, token, proxy, proxyPort)
+        {
+        }
+        internal virtual HttpWebRequest RequestStamping(byte[] xml, string version, string format, string operation, string email)
```
hace falta customId

RICH-CD replied:
Se Agrega customid

JuanGamezSW commented on SW-sdk/Services/Stamp/StampServiceV4.cs:
```diff
@@ -0,0 +1,37 @@
+            request.ContentType = "application/json";
+            request.Method = WebRequestMethods.Http.Post;
+            request.Headers.Add(HttpRequestHeader.Authorization.ToString(), "bearer " + this.Token);
+            request.Headers.Add("email", email);
```
Me parece que si fuera null email no se tendría que agregar.

RICH-CD replied:
Se valida que email sea distinto de null para agregarlo al request

JuanGamezSW commented on SW-sdk/Services/Stamp/BaseStampV4.cs:
```diff
@@ -0,0 +1,81 @@
+                return handler.HandleException(ex);
+            }
+        }
+        public virtual StampResponseV4 TimbrarV4(string xml, string email = null, bool isb64 = false)
```
falta parametro "customId" opcional.

RICH-CD replied:
Se agrega customid

JuanGamezSW commented on SW-sdk-45/Services/Stamp/StampServiceV4.cs:
```diff
@@ -0,0 +1,36 @@
+            this.SetupRequest();
+            Dictionary<string, string> headers = new Dictionary<string, string>() {
+                    { "Authorization", "bearer " + this.Token },
+                    { "email", email }
```
si es null no se tendría que agregar.
Faltaría tambien el customId.

RICH-CD replied:
Se agregan validaciones y customid

alxcodes commented:
revisar estas UT para que te orientes con las que te pueden faltar
https://github.com/lunasoft/SW-sdk-NetStandard/blob/master/test/sdk-test/Services/Stamp/StampV4_Test.cs

RICH-CD committed:
Se agrega customId al metodo de timbrado StampV4
Se agrega nueva clase UT al timbrado StampV4
Se agregan metodos UT al timbrado StampV4
